src/main/java/com/nvidia/spark/rapids/jni/kudo/SerializedTableHeader.java

```diff
@@ -0,0 +1,155 @@
+package com.nvidia.spark.rapids.jni.kudo;
+
+import java.io.DataInputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Optional;
+
+/**
+ * Holds the metadata about a serialized table. If this is being read from a stream
+ * isInitialized will return true if the metadata was read correctly from the stream.
+ * It will return false if an EOF was encountered at the beginning indicating that
+ * there was no data to be read.
+ */
+public final class SerializedTableHeader {
+    /**
+     * Magic number "KUDO" in ASCII.
+     */
+    private static final int SER_FORMAT_MAGIC_NUMBER = 0x4B55444F;
+    private static final short VERSION_NUMBER = 0x0001;
+
+    // Useful for reducing calculations in writing.
+    private long offset;
+    private long numRows;
+    private long validityBufferLen;
+    private long offsetBufferLen;
+    private long totalDataLen;
+    // This is used to indicate the validity buffer for the columns.
+    // 1 means that this column has validity data, 0 means it does not.
+    private byte[] hasValidityBuffer;
```

Comment:
This should be a bit mask, not byte-per-column.  We never have to bit shift the entire hasValidityBuffer mask or anything complex like that, just single-bit test this which is trivial to compute, both on the write and read sides, and this saves a lot of space when we're dealing with tons of columns and tons of partitions across tons of tasks.